src/twisted/trial/_dist/test/test_worker.py

```diff
@@ -86,197 +76,155 @@ class LocalWorkerAMPTests(TestCase):
-        self.assertEqual(failure.value, WorkerException("error"))
-        self.assertTrue(results)
-
-    def test_runErrorWithFrames(self):
-        """
-        L{LocalWorkerAMP._buildFailure} recreates the C{Failure.frames} from
-        the C{frames} argument passed to C{AddError}.
-        """
-        results = []
-        errorClass = fullyQualifiedName(ValueError)
-        d = self.worker.callRemote(
-            managercommands.AddError,
-            testName=self.testName,
-            error="error",
-            errorClass=errorClass,
-            frames=["file.py", "invalid code", "3"],
+        case = pyunitcases.PyUnitTest("test_error")
+        result = self.workerRunTest(case)
+        assert_that(
+            result,
+            matches_result(
+                # Failures don't compare nicely so unpack the result and peek inside.
+                errors=contains_exactly(
+                    isTuple(  # type: ignore[arg-type]
+                        equal_to(case),
+                        isFailure(
+                            type=equal_to(Exception),
+                            value=equal_to(WorkerException("pyunit error")),
+                            frames=has_item(  # type: ignore[arg-type]
+                                similarFrame("test_error", "pyunitcases.py")  # type: ignore[arg-type]
+                            ),
+                        ),
+                    ),
+                ),
+            ),
         )
-        d.addCallback(lambda result: results.append(result["success"]))
-        self.pumpTransports()
 
-        case, failure = self.result.errors[0]
-        self.assertEqual(self.testCase, case)
-        self.assertEqual(failure.type, ValueError)
-        self.assertEqual(failure.value, WorkerException("error"))
-        self.assertEqual([("file.py", "invalid code", 3, [], [])], failure.frames)
-        self.assertTrue(results)
-
-    def test_runFailure(self):
+    def test_runFailure(self) -> None:
         """
         Run a test, and fail.
         """
-        results = []
-        failClass = fullyQualifiedName(RuntimeError)
-        d = self.worker.callRemote(
-            managercommands.AddFailure,
-            testName=self.testName,
-            fail="fail",
-            failClass=failClass,
-            frames=[],
+        case = pyunitcases.PyUnitTest("test_fail")
+        result = self.workerRunTest(case)
+        assert_that(
+            result,
+            matches_result(
+                failures=contains_exactly(
+                    isTuple(  # type: ignore[arg-type]
+                        equal_to(case),
+                        isFailure(
+                            # AssertionError is the type raised by TestCase.fail
+                            type=equal_to(AssertionError),
+                            value=equal_to(WorkerException("pyunit failure")),
+                        ),
+                    ),
+                ),
+            ),
```

Comment:
I am not a big fan of these generic assertions.

we can leave this as it is, since `matched_result` is already in trunk.
I find the name a bit too generic.

Something to my taste.
```
assert_than(result, is_testrun_failure(case, type=AssertionError, value=WorkerException("pyunit failure")
```

and all the extra `isTuple`, `equal_to` and the general structure is handled by the dedicated matcher.

From my experience, with these specialiez markers it was easier to refactor the internal structure of a code as you might only need to update the specialized marker, without updating each test.

Reply:
This overlaps a bit with the comment above and I left a reply there.